### ISSUE TITLE
The Rendering Pipeline Discussion

### ISSUE THREAD

**Issue by oli414** [OpenRCT2/OpenGraphics#18]
# Introduction
Back in the day, RCT players would use sprite editing to create new scenery pieces, some users tried to do a bit more, like Amazing Earl. Currently one of the first user-generated sprite rendering pipelines that we know of. His initial efforts have shown that it is possible to create new graphics for the game similar to how Simon Foster would have done it.

OpenGraphics for OpenRCT2 was initially kickstarted by Zrowny in 2016 with a setup made for Blender 2.7-, creating the repository, and getting some of the first objects made. 

A while later in 2018, I tried to have my own shot at it, once again using Blender 2.7. Thanks to the groundwork by Zrowny, research from Spacek, X7 and Coors, and even some information from Simon Foster himself we got to a point where we had a very promising setup. Since then Blender 2.8 came out, which removed Blender Render. 

Luckily Frutiemax took it upon himself to show what we could do without Blender Render, creating his own set up along with several objects.

In the meantime, X7 created his very own renderer. Specifically designed to render vehicles and track sprites, some of which have already been added to the game.

# The Dilemma
Things were looking really good until Blender 2.8 came out. Blender 2.8 significantly improves Blender in its accessibility and has turned Blender into an industry tool. Unfortunately, to do so the developers had to remove Blender Render. Blender Render is a scanline renderer, just like the default scanline renderer in 3ds Max that Simon Foster used to render most of the RCT2 sprites. The only rendering methods currently left are Eevee, a hybrid optimized for fast rendering. Which utilizes OpenGL with ray tracing for more accurate lighting. And Cycles, a full-on path-tracing renderer. 

Frutiemax has done some amazing renders with Cycles. Unfortunately, not every aspect has been quite figured out yet which may cause complications.

X7 on the other hand has been working on his own renderer which is highly optimized for track rendering. The results are really solid and a custom renderer would allow us to be independent of what rendering engines Blender contains. Maintaining a custom renderer has its own challenges.

# What we're discussing
As it currently stands the dilemma has meant that we're not able to move forward with the OpenGraphics project. We've had countless discussions, but we can't come to a conclusion without carefully considering all our options. Allowing all the major contributors to chip in. And provide evidence for how different options might compare.

The main question is: What rendering engine will we use?

From there we'll have other aspects to discuss regarding the pipeline. The rendering engine is the biggest roadblock at the moment that we need to get out of the way before we can move on.

# The options
Currently, we have the following options to consider:
- Use Blender 2.8+ with the Cycles rendering engine
- Use Blender 2.8+ with the Eevee rendering engine
- Use Blender 2.7- with the Blender Render rendering engine
- Develop and use a standalone custom renderer
- Develop and use a custom renderer for Blender 2.8+

Please share and discuss thoughts and concerns for each option.

**Comment by IntelOrca** [comment]
My opinion may have less importance since I am not an object modeler.

> Use Blender 2.7- with the Blender Render rendering engine

I would personally rule this out. Relying on an old version of Blender means you'll miss out on any new feature or compatibility they add later. I don't know how possible it is to export from newer versions back to 2.7, but it seems like unnecessary hassle.

This leaves deciding between whether we use our own renderer or Blender's.

I like the idea of a custom renderer because we theoretically have complete freedom over it, not just how it looks, but how we can get our models in, and an asset out, ready for OpenRCT2 (particularly when it comes to using the correct palette indices). It also means Blender will never be able to break, or change the way our objects look. I imagine this could also be easier to work with when it comes to automated scripts and CI.

Question is: how difficult it is to get a custom renderer to do all the features you like from Blender's renderers?

> Develop and use a custom renderer for Blender 2.8+

If we use a custom renderer, I would have thought we'd want it to be useable outside Blender, but still have a plugin for Blender that uses it, so that modelers can get immediate previews of what their object will look like.

**Comment by oli414** [comment]
I've been asked to just make the call before but I really don't want to without giving everyone a chance to chip in.

My personal preference would be Blender 2.7. Even though I'm using Blender 2.8+ on a daily basis, Blender 2.7 can be a pain to work with after getting used to Blender 2.8+. But it has proven itself very well, and does supports we need.

# A custom solution
First of all, I don't think creating our own renderer is an option. Currently, X7 has done an amazing job creating his own renderer, and it could be a potential candidate. But as it currently stands X7 is the only person that knows how it works. And I'm very hesitant to maintain for such a complex and low-level tool in a sub-project that will mostly see people with creative skills.

In addition to that, I think that a custom renderer would need so many features to cover everything we need, that it would out of scope for this project. Think of things like lighting, reflectivity, glass, and many more features that have to be supported, especially in order to work in conjunction with a tool like Blender.

In my experience creating a custom renderer that is feature complete is a massive task. To give a cheesy anecdote: In my final year of uni a group of very knowledgeable graphics programmers that now work for companies like Guerrilla Games and Massive were unable to complete a custom rendering engine for a 3d modeling engine within half a year. Working on it full-time.

My recommendation would be to not use a custom renderer (apart from some specific things perhaps like track rendering) I believe that it would be more than we can take on, especially whilst there are other already working solutions available.

# Blender 2.8
I've done a lot of tests with Blender 2.8 with both Cycles and Eevee. But due to the nature of those rendering engines trying to replicate real lighting I have been struggling to get results that were precise enough to work for isometric sprites that may need to be tileable.

Additionally, I was having a lot of trouble to control the way textures show up. Ideally, you can control the texture filtering and anti-aliasing independently but I found it to be nearly impossible to get noisy textures to remain noisy while also having anti-aliasing enabled to smooth out the geometry itself.

# Blender 2.7
The only reason I wouldn't recommend Blender 2.7 is that its user interface is horrendous once you've gotten used to Blender 2.8, or come from another software package.

Apart from that, it has proven itself to be capable of doing everything it has to. The way lighting is calculated is very predictable which makes creating tileable objects easy. Additionally, you have a lot of control over the texture filtering, with several great texture filtering algorithms available. Even the filter size is adjustable independently from the Anti-aliasing filter size.

**Comment by Ryder17z** [comment]
Well, here's my take on this:

> ### Blender
> The only reason I wouldn't recommend Blender 2.7 is that its user interface is horrendous once you've gotten used to Blender 2.8, or come from another software package.

I come from the AutoCAD/Inventor/Solidworks/etc side and Blender 2.7 is a horrid transition for me. I've only just started picking up Blender 2.8x as it's a literal day/night difference.

Blender 2.7 isn't bad, but it is difficult to get into unless you learned it first. 
As mentioned previously, there are issues exporting from newer versions of blender to 2.7x so I believe transitioning to 2.8x would be for the best.

> ### A custom solution
Here is a thought: It could partially rely on some version of Blender.

If it comes down to sticking with blender 2.7, then I would like to see a similar process to this for those who wish to use Blender 2.8x:

1. create model and texture it
2. export as 3ds/something which has the materials or textures properly embedded
3. import object by a custom script in legacy Blender which applies fixes that might be required.
4. run the render

* I have not attempted to export anything from blender to a different piece or software or to an older version of blender so i don't know what issues are common there. I do know that texture/material handling and scaling/rotation can be iffy in general though (I often see it going between AutoCAD/Inventor and Solidworks)

**Comment by frutiemax** [comment]
> 1. create model and texture it
> 2. export as 3ds/something which has the materials or textures properly embedded
> 3. import object by a custom script in legacy Blender which applies fixes that might be required.
> 4. run the render

The problem with this is that there are some cases where you need particle systems or scenes with a lot of objects. The import process can become quite time consuming. You can also have objects with sculpting and these are not easily exported to 3ds/fbx.

I share the same opinion as Ryder17z and also IntelOrca. Eevee is supposed to be a replacement to the blender internal renderer, so perhaps there is some experimentation to be done in that aspect.

I had also some good looking results with Cycles, but in some aspect, the blender internal gives a more old school render style when I tested some tree renders. The blender 2.7 interface was a bit frustrating to work with (right-click to select object???) and not being able to view the material shader in real time kind of sucks. But I could eventually achieve what I wanted in the end...

**Comment by oli414** [comment]
You could always try to model something in 2.8 first and then redo your materials manually in 2.7. I don't even think you'd need a tool for that. That's probably a process I'd personally consider. though on the other hand, 2.7 was fine once you got used to it before 2.8 so it would just require a bit of getting used to again. 2.8 spoiled us a bit haha

> right-click to select object
You can luckily change that in the settings :-P

**Comment by oli414** [comment]
> I would personally rule this out. Relying on an old version of Blender means you'll miss out on any new feature or compatibility they add later. I don't know how possible it is to export from newer versions back to 2.7, but it seems like unnecessary hassle.

I'm not sure why we would need new features. The original graphics were made with an old version of 3ds max, any QOL improvements we get in 2.7 are nice to have but I don't think we really need more than that. The graphics are pretty basic after all. Compatibility could be a concern further down the road but I can't imagine that to become relevant in the coming 10 years.

**Comment by nfeho** [comment]
**Blender**
From the above mentioned options, I tried only rendering the assets in Blender 2.8 cycles. While this renderer was ideal for me so far, I have only tried rendering plants and small trees using particle systems. I understand that for majority of objects, the scanline renderer is probably more useful and produces better results, which fit more with RCT2 style. I am not sure about the look of trees or other particle based objects in renderers other than cycles.

**Custom renderer**
Developing a standalone renderer is really hard, and as Oli pointed out, it could be a huge time-eating project, with no guarantee that the resulting sprites would look much more authentic. On topic of custom renderers, X7 has already made one and used it to create the ride tracks, that were merged into main project. I have yet to see how the scenery pieces would look like being rendered with his setup. Maybe we should look into rendering more types of objects with his setup and compare the results to other renderers.

**Multiple renderers**
We could use multiple renderers for creating assets. I know, that the main focus of this issue is to unify the asset creation and choose one of the options, but we could manage 2 setups, with each being specialized for different sets of sprites. For example we could use X7's renderer for rendering tracked rides and use whichever Blender renderer we agree on for rendering pretty much everything else.

**Comment by oli414** [comment]
Alright so one of my biggest concern in the 2.8 vs 2.7 discussion is the lighting, crispness and texture quality.

I think the latter is a pretty big thing for me. Getting textures to show up in a crisp manner, and not blurry has been a bit of a challenge for me when using 2.8. I can get it to work, but only if I fully ignore anti-aliasing.

Anti-aliasing in 2.8 is the result of the samples that are taken, and the size of the region within which the samples are taken. But this directly influences the way textures are filtered as well.

You do get the following setting on the Image Texture node:
![image](https://user-images.githubusercontent.com/2348094/121715468-37588d00-cadf-11eb-8560-bb7cbcbfc5c6.png)
But in my experience, this has very little effect. As the previously mentioned issue still applies.

In 2.7 we get the following options to set the way the texture is interpolated:
![image](https://user-images.githubusercontent.com/2348094/121715888-ac2bc700-cadf-11eb-9ff0-1230af3c2854.png)

The texture is still influenced by the anti-aliasing:
![image](https://user-images.githubusercontent.com/2348094/121715936-bbab1000-cadf-11eb-9069-610c69e04ba5.png)
But you can still get pretty good results.

Allow me to show some tests that I've done:

I'm rendering a 1x1 (tile) cube with the following texture:
![Bricks](https://user-images.githubusercontent.com/2348094/121721702-d2a03100-cae4-11eb-9c2a-a20772f0619e.jpg)

## Blender 2.8
### No AA
Film Pixel Filter: Blackman-Harris 0.01 px
Linear texture filtering:
![linear_noaa](https://user-images.githubusercontent.com/2348094/121721554-ac7a9100-cae4-11eb-84c8-22fe80e2a799.png)
Closest texture filtering:
![closest_noaa](https://user-images.githubusercontent.com/2348094/121721574-b2707200-cae4-11eb-87c6-4bfee94ebaeb.png)
Cubic texture filtering:
![cubic_noaa](https://user-images.githubusercontent.com/2348094/121721594-b8665300-cae4-11eb-922f-286f91545862.png)

### With AA
Film Pixel Filter: Blackman-Harris 1.4 px
Linear texture filtering:
![linear_aa](https://user-images.githubusercontent.com/2348094/121722098-4e01e280-cae5-11eb-8cc0-f1a061330f2c.png)
Closest texture filtering:
![closest_aa](https://user-images.githubusercontent.com/2348094/121722110-51956980-cae5-11eb-8841-9b818893ce8a.png)
Cubic texture filtering:
![cubic_aa](https://user-images.githubusercontent.com/2348094/121722089-4b06f200-cae5-11eb-94c8-ebfb4aec2595.png)

## Blender 2.7
### No AA
Anti-Aliasing: Off
Feline Texture Filtering 0.1:
![feline1_noaa](https://user-images.githubusercontent.com/2348094/121723099-66263180-cae6-11eb-8dbe-2ac49e8cbfa6.png)
EWA Texture Filtering 0.4:
![ewa4_noaa](https://user-images.githubusercontent.com/2348094/121723445-cf0da980-cae6-11eb-89d4-35aae04b2aa9.png)
Area Texture Filtering 0.4:
![area4_noaa](https://user-images.githubusercontent.com/2348094/121722980-442caf00-cae6-11eb-92d4-326d5b2f10b7.png)

### With AA
Anti-Aliasing: 8 Samples, Box 1.4 px

Feline Texture Filtering 0.1:
![feline1_aa](https://user-images.githubusercontent.com/2348094/121723082-60305080-cae6-11eb-8a70-80cdc83089d2.png)
EWA Texture Filtering 0.7:
![ewa7_aa](https://user-images.githubusercontent.com/2348094/121723435-cb7a2280-cae6-11eb-82e4-2b447427960f.png)
Area Texture Filtering 0.8:
![area8_aa](https://user-images.githubusercontent.com/2348094/121722961-40009180-cae6-11eb-8bd1-bf97f1526227.png)

As you can see, thanks to the options available in 2.7 I'm able to maintain the look of the texture, regardless of the anti-aliasing. The reason that this is important is that anti-aliasing isn't used consistently throughout the sprites in RCT.
Wall pieces often have it turned off to make them tileable, where something like a statue usually does anti-aliasing to make them stand out more.
A piece that doesn't use anti-aliasing may be used together with a piece that does have anti-aliasing. If these pieces share the same texture, you'd want them to appear the same in the sprite.

I'd love to see if someone else is able to have more control over their textures in Blender 2.8, let me know if you find something.

**Comment by RetroNuva10** [comment]
> 
> > Use Blender 2.7- with the Blender Render rendering engine
> 
> I would personally rule this out. Relying on an old version of Blender means you'll miss out on any new feature or compatibility they add later. I don't know how possible it is to export from newer versions back to 2.7, but it seems like unnecessary hassle.
> 
Very few of the new features presented with Blender 2.8 would ever actually be used within the scope of this project - most of the features have no effect on the actual modeling workflow, besides materials being applied differently (and arguably more time-consuming). I'm personally fluent in Blender 2.7 and have made several models using Oli's rendering plugin and, after getting a few things set up beforehand, it works like a charm, and the results speak for themself:

![ovenShowcase](https://user-images.githubusercontent.com/61289806/124505080-b9626b80-dd96-11eb-8644-ff2fe91af615.png)
![thumb](https://user-images.githubusercontent.com/61289806/124505083-bbc4c580-dd96-11eb-93c0-68164fd8a9ca.png)
![ovenBig](https://user-images.githubusercontent.com/61289806/124505085-bd8e8900-dd96-11eb-9bf5-e894a3a7fcaa.png)
![restroomShowcase](https://user-images.githubusercontent.com/61289806/124505087-bf584c80-dd96-11eb-8072-0b49d7b047b2.png)

Each of these renders is the raw output of Oli's plugin, with no touchup whatsoever. As Oli explained above, there are a couple parameters you'd want to adjust for any materials you use for the best results, but considering most objects would only use a dozen or fewer materials, it ends up being worth it.

Ultimately, my vote goes to using Blender 2.7. I understand that the UI is not very user-friendly, but I learned the basic ropes of Blender 2.7 about 4 years ago in under two hours - I'm more than happy to teach/give assistance to anyone else if they're willing to use 2.7 for Open Graphics. Once you get a grip on the hotkeys, you're good to go.

**Comment by oli414** [comment]
I agree with that, in addition Zrowny has made some wonderful improvements to my initial setup which paves the way for a fully featured setup.

I personally find the provided arguments against 2.7 to be lacking, from what I can gather it mostly comes down to:
- Less user friendly UI (Just a matter of getting used to it again)
- Unable to achieve new features and compatabilities (Since we're trying to recreate old graphics I doubt we need that for anything but convenience)

Keep in mind that we're not stuck with the choice we make now. As long as we aim for the look of the original graphics we can try to do so with a different renderer in the future if we ever feel the need to switch. There are some changes in tools used between RCT1 and RCT2 that aren't too noticeable either.

**Comment by Ryder17z** [comment]
Really unfortunate because that cuts some users off. For me it's not a matter of "just getting used to it again". I've just begun learning 2.8 because it's actually somewhat sensible from where i come from (CAD).

**Comment by RetroNuva10** [comment]
> Really unfortunate because that cuts some users off. For me it's not a matter of "just getting used to it again". I've just begun learning 2.8 because it's actually somewhat sensible compared to 2.7, factoring in where I'm coming from (CAD).

No matter which option we decide on, there will always be some given number of people who will have to learn the chosen tool, whether it be Blender 2.7, Blender 2.8, or some custom renderer. Thus, it boils down to whichever is the best option in terms of availability and quality of results. Like I said - I'm available frequently on Discord and am more than happy to help others get accustomed to 2.7. Alternatively, there are ways you could model in 2.8 and then transfer it over to 2.7 just for the rendering portion, if you really want to.

**Comment by Ryder17z** [comment]
>  Alternatively, there are ways you could model in 2.8 and then transfer it over to 2.7 just for the rendering portion, if you really want to.
That needs to be looked into but it would suffice.

**Comment by oli414** [comment]
Thanks for chiming in everyone. I think that'll mark the end of the dreaded version discussion :-P.

We'll base our tool on Blender 2.7 and use that for most of our renders for the foreseeable future. I'm incredibly excited that we can finally start setting some stuff in stone.
